ESQL: Clone ql for esql

alex-spies opened the pull request:
Part of https://github.com/elastic/elasticsearch/issues/106679

* Copy the `ql` project into a different project _just for esql_, call it `esql-core`.
* Make `esql` depend only on the latter.
* Fix `EsqlNodeSubclassTests`; I'm confused why this didn't bite us earlier.
* Update the warning regexes in some csv tests as the exceptions have other package names now.

**Note to reviewers:** Exclude the first commit when viewing the diff, as that contains only the actual copying of `ql`. The remaining commits are the actually meaningful ones. _The `build.gradle` files probably require the most attention._

alex-spies committed:
Use a clone of ql in esql, esql-core

alex-spies committed:
Update build.gradle files

alex-spies committed:
Add readme

alex-spies committed:
Update benchmark deps

alex-spies committed:
Fix EsqlNodeSubclassTests

alex-spies committed:
Fix multi-cluster-search-security

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

alex-spies commented:
Also kudos to @luigidellaquila - he made the first attempt in https://github.com/elastic/elasticsearch/pull/106079, which however included renaming the namespaces and fixing up any imports.

luigidellaquila commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/tree/EsqlNodeSubclassTests.java:
```diff
@@ -142,8 +142,8 @@ Expression randomResolvedExpression(Class<?> argClass) throws Exception {
-                var candidate = randomFrom(subclassesOf(asNodeSubclass));
-                if (UNRESOLVED_CLASSES.contains(candidate) == false) {
+                var candidate = randomFrom(subclassesOf(asNodeSubclass, CLASSNAME_FILTER));
+                if (UNRESOLVED_CLASSES.stream().allMatch(unresolved -> unresolved.isAssignableFrom(candidate) == false)) {
```
Good catch!

> I'm confused why this didn't bite us earlier.

probably just because the only missing case was `UnresolvedNamePattern`, that is a recent addition
At this point I think we can remove `UnresolvedAlias` and `UnresolvedStar` from the list, that are subclasses of `UnresolvedNamedExpression`

luigidellaquila reviewed (commented):
Thanks Alex!
Just a couple of observations

This needs to be changed probably

https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ml/qa/native-multi-node-tests/build.gradle#L18

It was not in my PR because it was added just one week after that.


This mixes EQL and ESQL, not sure if it can lead to classpath issues

https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/qa/consistency-checks/build.gradle#L16

luigidellaquila commented on x-pack/plugin/esql-core/build.gradle:
```diff
@@ -0,0 +1,24 @@
+esplugin {
+  name 'x-pack-esql-core'
+  description 'Elasticsearch infrastructure plugin for ESQL'
+  classname 'org.elasticsearch.xpack.ql.plugin.QlPlugin'
```
I'd change the plugin class (simple) name at least. Eg something like `QlCorePlugin`

rjernst replied:
Rene is right, this needs to be a different package.

alex-spies replied:
Moved that to `xpack.plugin.esql.core`.

costin reviewed (approved):
Since we don't aim to share the QL clone outside ESQL, why use it as a separate module instead of moving it directly into ESQL?

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into clone-ql-for-esql

alex-spies committed:
Address Luigi's remarks

alex-spies committed:
Commit Luigi's ql migration script

Co-authored-by: Luigi Dell'Aquila <luigi.dellaquila@elastic.co>

alex-spies committed:
Update Luigi's script

alex-spies committed:
Rename package ql -> esql.core

alex-spies committed:
Fix bwc tests

alex-spies committed:
Remove Luigi's script

alex-spies commented:
> Since we don't aim to share the QL clone outside ESQL, why use it as a separate module instead of moving it directly into ESQL?

Moving that into the `esql` plugin directly would be more elegant. We could do that in a follow up if desired; I'd prefer not to do it in this PR, as that change is a bit less mechanical than just copying all the things, so more could go wrong.

alex-spies committed:
Spotless + checkStyle

alex-spies committed:
More checkStyle

luigidellaquila commented:
> Moving that into the esql plugin directly would be more elegant. We could do that in a follow up if desired; I'd prefer not to do it in this PR, as that change is a bit less mechanical than just copying all the things, so more could go wrong.

It could be less obvious than it seems: today Compute depends on QL, and we don't want it to depend on ESQL

alex-spies committed:
Update esql-core's README.md

alex-spies commented:
Thanks for the reviews @costin , @luigidellaquila , @breskeby and @rjernst !

I addressed your remarks, all should be good now. In particular, moved the `esql-core` classes into an own package, `xpack.plugin.esql.core`.

We can move the contents of `esql-core` into just `esql` later - hopefully placing the classes in `esql.core` will mean we won't have to update all the imports _yet another time_ then.

luigidellaquila reviewed (approved):
Still a couple of fixes to make the CI happy ([MlNativeIntegTestCase](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ml/qa/native-multi-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/ml/integration/MlNativeIntegTestCase.java#L162) and [EvalBenchmark](https://github.com/elastic/elasticsearch/blob/main/benchmarks/src/main/java/org/elasticsearch/benchmark/compute/operator/EvalBenchmark.java#L34-L39)), but apart from that, LGTM

Thanks Alex!!!

As a follow-up (this PR is big enough already), we'll have to double-check that the code generator still respects the code style (eg. the order of imports)

alex-spies committed:
Update forbidden signatures

alex-spies committed:
Fix more bwc tests

alex-spies committed:
Fix benchmark and MlNativeIntegTestCase

alex-spies committed:
Fix dependency esql -> esql-core

This cannot be compileOnly; otherwise some tests fail because the
SingleValueQuery class cannot be loaded.

alex-spies committed:
Revert "Fix dependency esql -> esql-core"

This reverts commit b4d92bd7c4cf082d565e07dbbf2f58ac49fef36d.

I was wrong :D

alex-spies committed:
Fix CrossClusterShardTests

These need `esql-core`. Previously, these depended on `ql` and we
somehow didn't see them fail, maybe because `eql` and `sql` had the
class loader load the required classes.

alex-spies committed:
Fix some more bwc tests

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into clone-ql-for-esql

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into clone-ql-for-esql
